operators/src/pro/cache/shared_cache.rs

```diff
@@ -0,0 +1,1190 @@
+    fn evict_entries_until_can_fit_bytes(&mut self, bytes: usize) {
+        self.evict_until_can_fit_bytes(bytes);
+    }
+}
+
+pub trait CacheView<C, L>: CacheEvictUntilFit {
+    fn operator_caches_mut(
+        &mut self,
+    ) -> &mut HashMap<CanonicOperatorName, OperatorCacheEntry<C, L>>;
+
+    fn create_operator_cache_if_needed(&mut self, key: CanonicOperatorName) {
+        self.operator_caches_mut()
+            .entry(key)
+            .or_insert_with(|| OperatorCacheEntry::new());
+    }
+
+    fn remove_operator_cache(
+        &mut self,
+        key: &CanonicOperatorName,
+    ) -> Option<OperatorCacheEntry<C, L>> {
+        // TODO: maybe remove the size of the OperatorCacheEntry to the cache size?
+        self.operator_caches_mut().remove(key)
+    }
+}
+
+#[allow(clippy::type_complexity)]
+pub struct OperatorCacheEntryView<'a, C: CacheElement> {
+    operator_cache: &'a mut OperatorCacheEntry<
+        CacheQueryEntry<C::Query, C::CacheContainer>,
+        CacheQueryEntry<C::Query, C::LandingZoneContainer>,
+    >,
+    cache_size: &'a mut CacheSize,
+    landing_zone_size: &'a mut CacheSize,
+    lru: &'a mut LruCache<CacheEntryId, TypedCanonicOperatorName>,
+}
+
+impl<'a, C> OperatorCacheEntryView<'a, C>
+where
+    C: CacheElement + ByteSize,
+    C::Query: Clone + CacheQueryMatch,
+    CacheQueryEntry<C::Query, C::LandingZoneContainer>: ByteSize,
+    CacheQueryEntry<C::Query, C::CacheContainer>: ByteSize,
+{
+    fn is_empty(&self) -> bool {
+        self.operator_cache.is_empty()
+    }
+
+    /// This method removes a query from the landing zone.
+    ///
+    /// If the query is not in the landing zone, this method returns None.
+    ///
+    fn remove_query_from_landing_zone(
+        &mut self,
+        query_id: &QueryId,
+    ) -> Option<CacheQueryEntry<C::Query, C::LandingZoneContainer>> {
+        if let Some(entry) = self.operator_cache.remove_landing_zone_entry(query_id) {
+            self.landing_zone_size.remove_element_bytes(query_id);
+            self.landing_zone_size.remove_element_bytes(&entry);
+
+            // debug output
+            log::debug!(
+                "Removed query {}. Landing zone size: {}. Landing zone size used: {}, Landing zone used percentage: {}.",
+                query_id, self.landing_zone_size.total_byte_size(), self.landing_zone_size.byte_size_used(), self.landing_zone_size.size_used_fraction()
+            );
+
+            Some(entry)
+        } else {
+            None
+        }
+    }
+
+    /// This method removes a query from the cache and the LRU.
+    /// It will remove a queries cache entry from the cache and the LRU.
+    ///
+    /// If the query is not in the cache, this method returns None.
+    ///
+    fn remove_query_from_cache_and_lru(
+        &mut self,
+        cache_entry_id: &CacheEntryId,
+    ) -> Option<CacheQueryEntry<C::Query, C::CacheContainer>> {
+        if let Some(entry) = self.operator_cache.remove_cache_entry(cache_entry_id) {
```

Comment:
can't you re-use the code in/from `evict_until_can_fit_bytes`?

Reply:
i don't see how to do that in a nice way. The problem is that this is a different struct and in the evict method we have to look at the typed names.